### ISSUE TITLE
How to combine Ajax abort with eventChannel?

### ISSUE THREAD

**Issue by zushenyan** [redux-saga/redux-saga#865]
I was expecting `eventChannel` to emit an `abort` event in an Ajax while listening to various events like `progress`, but it always emit `progress` event when I was trying to abort.

Here is the event log:
![redux-saga](https://cloud.githubusercontent.com/assets/829686/23911906/29d5b27a-0919-11e7-8af2-9431fe0a96d6.png)

Here is my code in `client/redux-saga/sagas.js`
```js
import { call, put, take, select } from "redux-saga/effects";
import { eventChannel, END } from "redux-saga";
import { createAction } from "../utils.js";
import * as actions from "../actions.js";
import * as api from "../api.js";

export const FETCH_DATA_SAGA       = "fetch data saga";
export const FETCH_DATA_ABORT_SAGA = "fetch data abort saga";
export const fetchData             = createAction(FETCH_DATA_SAGA);
export const fetchDataAbort        = createAction(FETCH_DATA_ABORT_SAGA);

export const createFetchDataChannel = () => {
  return eventChannel(emit => {
    const xhr = api.fetchData({
      onLoadStart: (e) => emit(actions.fetchDataStart({xhr})),
      onLoad:      (e) => {
        emit(actions.fetchDataDone({data: e.target.response, xhr}));
        emit(END);
      },
      onProgress:  (e) => emit(actions.fetchDataLoading({data: e.target.response, xhr})),
      onError:     (e) => {
        emit(actions.fetchDataFail({error: e.target.response, xhr}));
        emit(END);
      },
      onAbort:     (e) => {
        emit(actions.fetchDataAbort({xhr}));
        emit(END);
      }
    });
    xhr.send();
    return () => {};
  });
};

export function * fetchDataSaga() {
  try{
    const channel = yield call(createFetchDataChannel);
    while(true){
      const action = yield take(channel);
      yield put(action);
    }
  }
  catch(e){
    console.log(e);
  }
}

export function * fetchDataAbortSaga() {
  const xhr = yield select(state => state.xhr);
  xhr && xhr.abort();
}

export function * watchFetchDataSaga() {
  while(true){
    yield take(FETCH_DATA_SAGA);
    yield call(fetchDataSaga);
  }
}

export function * watchFetchDataAbortSaga() {
  while(true){
    yield take(FETCH_DATA_ABORT_SAGA);
    yield call(fetchDataAbortSaga);
  }
}

export default function * rootSaga(){
  yield [
    watchFetchDataSaga(),
    watchFetchDataAbortSaga()
  ];
}
```

Here is the one I was expecting: (This is written in `redux-observable`)
![redux-observable](https://cloud.githubusercontent.com/assets/829686/23911953/4fb57cc8-0919-11e7-80c5-e51cb78159cc.png)

In my `redux-saga` example, though the code in 
```js
onAbort: (e) => {
  emit(actions.fetchDataAbort({xhr}));
  emit(END);
}
```
really get executed, but the last action being dispatched (`put` in `redux-saga`) is always this one
```js
onProgress:  (e) => emit(actions.fetchDataLoading({data: e.target.response, xhr})),
```

My question is **how do I make my `redux-saga` example result being like `redux-observable` example? Did I do something wrong with `eventChannel`?**

The full code is being placed [here](https://github.com/zushenyan/redux-comparison), which is an comparison between `redux-thunk`, `redux-saga` and `redux-observable` when it comes to handling a little complicated situations like responding `progress` or `abort` events in Ajax.

**Comment by Andarist** [comment]
Are u sure its not working? Ive managed to see it working just like you would have it expected to, just had to throttle my network a little bit, as this file was downloading too fast and I wasnt able to click cancel fast enough.

**Comment by zushenyan** [comment]
Thanks for reply. 

I am sure it is not working... In my first pic I did throttle my network to `Good 2G` and then click cancel for cancellation but still no luck. Are you sure you were running in `http://localhost:8082/redux-saga`?

In example of `redux-observable`, if you abort an request, it will display `abort` text on the screen because the last action is `fetchDataAbort`, however in `redux-saga` the last action will be`fetchDataLoading` and still display `loading` on screen, which means not working as intended.

For some unknown reasons ` const action = yield take(channel);` in `fetchDataSaga` didn't take `fetchDataAbort` action even though this one

```js
onAbort: (e) => {
  emit(actions.fetchDataAbort({xhr}));
  emit(END);
}
```
really get executed.

**Comment by Andarist** [comment]
yes, im was running this on correct url, r u sure you have refreshed ur tab after setting up throttling? i think it might kick in only after refresh, however im not 100% sure

we'll try to test it one more some time later though

**Comment by zushenyan** [comment]
I found a solution for this issue. It's making codes in `onAbort` run in deferred

```js
onAbort: (e) => {
  setTimeout(() => {
    emit(actions.fetchDataAbort({xhr}));
    emit(END);
  }, 0);
}
```

Here is the result I want
![screen shot 2017-03-15 at 11 58 41 pm](https://cloud.githubusercontent.com/assets/829686/23957945/9c7a5764-09db-11e7-8fea-956825357c95.png)

However I don't know why it works and why it also works without `deferred` in your environment. 🤔

**Comment by Andarist** [comment]
Sorry for not responding any time sooner, frankly I missed your reply somehow. I guess you have worked around your issue by now, but if you have any more questions or your issue stays unresolved - please just write in this thread and I will for sure get back to you this time.
